PHP 8.0, 8.1, PHPUnit 9 support + automated check for php compatibility

Zombaya opened the pull request:
I've expanded on the excellent work done by @27pchrisl  in PR #2332.  I rebased his branch onto the latest version of master, upgraded the new tests to be able to use phpunit 9 and added a new CI-job that lints all code in `src` to make sure it is compatible with all supported versions of php (5.5-8.1) using [phpcompatibility](https://github.com/PHPCompatibility/PHPCompatibility).

This should add support for php 8.0 and php 8.1 to the project and make sure no code will trigger bugs because of changes in php.

This should probably fix #2378, #2367

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

stobrien89 commented:
Hi @Zombaya and @27pchrisl,

Thanks a lot for all of the work on this.  @SamRemis and I will review when we have a chance, which will take some time. We'll get back to you ASAP, likely by the end of next week.

Zombaya commented:
Very much looking forward to it :-).  If I can be of any help, please let me know and I'll try to take some time to help.

SamRemis commented on tests/Polyfill/PHPUnit4/PHPUnitCompatTrait.php:
```diff
@@ -0,0 +1,131 @@
+        $this->assertNotContains($expected, $actual, $message);
+    }
+
+    public function assertStringContainsStringIgnoringCase($expected, $actual, $message = '')
```
Same as above assertStringContainsIgnoreCase()

27pchrisl commented on src/MockHandler.php:
```diff
@@ -37,7 +37,7 @@ public function __construct(
 
         if ($resultOrQueue) {
-            call_user_func_array([$this, 'append'], $resultOrQueue);
+            call_user_func_array([$this, 'append'], array_values($resultOrQueue));
```
This is in place for the PHP 8 named params change, where in PHP7 for call_user_func_array() array keys were ignored. Now, string keys will be interpreted as parameter names. The simple solution is to strip keys using array_values.

27pchrisl commented on tests/Polyfill/PHPUnit4/PHPUnitCompatTrait.php:
```diff
@@ -0,0 +1,131 @@
+        $this->assertContains($expected, $actual, $message);
+    }
+
+    public function assertStringNotContainsString($expected, $actual, $message = '')
```
Agreed it is long! But it's new in phpunit (https://phpunit.readthedocs.io/en/9.5/assertions.html#assertstringcontainsstring). And PhpStorm includes a linter rule to hint it should be used. However, it is certainly dealer's choice so we could revert this.

27pchrisl commented on src/Crypto/AesGcmDecryptingStream.php:
```diff
@@ -83,6 +83,7 @@ public function createStream()
                 $this->keySize
             ));
         } else {
+            // phpcs:ignore PHPCompatibility.FunctionUse.NewFunctionParameters -- False positive, never used in php <= 7.0
```
In my loosely held opinion in the tests is best, as then your rules and your tests are right next to each other in your editor view.

Zombaya replied:
I agree that using the comments is rather ugly but it's the only way to ignore specific sections of code.

Maybe we could go for a middleground where we have multiple configuration-files:
* All rules, all files except the ones with exceptions
* All rules except false positive rule 1, single file which had the false positive
* All rules except false positive rule 2, single file which had the false positive
* ...

This way, we would ignore the false positive in a whole file, which might lead to a false negative in case a rule would trigger twice in a file with one false positive.

This would allow to have the bulk of the files linted with all rules and files with false positives would only have one rule ignored.

Zombaya commented on .github/workflows/tests.yml:
```diff
@@ -39,34 +39,48 @@ jobs:
         vendor/bin/phpstan analyse src
 
+    #static compatibility-check for all versions
+    #Only needs to be run on 1 php-version to check for all versions
```
If there is 100% code coverage, this will be redundant, yes.  

When I added the tool, only false positives were detected, so in that regard this tool does not add much value to the project as it stands.

However, I think the output from the tool is still useful in case some new code is using non-bc-compatible syntax or function-arguments.  I think this output might be more clear than a failing unittest, where you as a developer need to be aware what exactly is not valid in all versions of php.

I've used this tool in other projects, which did have non-compatible code.  There it helped to figure out which code needed some refactoring.

If you decide the tool does not bring much value, you can simply drop 7c51a9ea79fd6dc435dd9e2b0d1e94b0743ba092 from the PR as it was added in a seperate commit.  Worst case my effort lead to a validation that all code is valid in php 5.5-8.1 :slightly_smiling_face:.

stobrien89 commented:
Hey @Zombaya,

I know it's been a while— sorry for the delay.  Could you rebase and remove the compatibility check? We've decided it's something we'll possibly revisit in the future.  Thanks again for all of the effort on this.

Zombaya commented:
Hey @stobrien89,

I've rebased everything again, added some small fixes because of changed tests and removed the commit about php-compatibility-check.

I ran the check manually once more and it succeeded without problems.

If you ever want to re-integrate the tool, I made [a backup of the branch](https://github.com/Zombaya/aws-sdk-php/tree/phpcompatibilityChecker-backup) or you could always check [the commit itself](https://github.com/Zombaya/aws-sdk-php/commit/a83f6aa57746b3f5c4e48c6d4d0e1355918602bf).

Zombaya commented on tests/Crypto/Polyfill/AesGcmTest.php:
```diff
@@ -2618,9 +2618,8 @@ public function testCompat()
         }
-        $ptLen = \random_int(0, 1024);
-        $aadLen = \random_int(0, 1024);
-        $i = 0;
+        $ptLen = \random_int(1, 1024);
+        $aadLen = \random_int(1, 1024);
```
The reason the pipeline for this MR previously failed, was because `$aadLen` was 0, resulting in a failure when doing `$aad = \random_bytes($aadLen + $i);`.

This is now avoided by having the minimum-value of both lengths to be 1.

Zombaya commented:
@stobrien89 , I'm finished for today.  I believe all warnings from php 8.1 about deprecated ways of using some functions or implicit type-conversions should be fixed, as well as deprecation-warnings from phpunit 9.x and I solved the very low possibility of a failed workflow because of generating an invalid stringlength.

I believe this MR is now ready again.  If you need any more changes or information, I'll be glad to help.

Zombaya commented on src/Credentials/AssumeRoleWithWebIdentityCredentialProvider.php:
```diff
@@ -114,7 +114,7 @@ public function __invoke()
                     if (empty($token)) {
                         if ($this->tokenFileReadAttempts < $this->retries) {
-                            sleep(pow(1.2, $this->tokenFileReadAttempts));
+                            sleep((int) pow(1.2, $this->tokenFileReadAttempts));
```
I've kept the result of the code the same, but it does lose some precision which might be wanted.

An alternative fix would be using `usleep()`, to keep a higher resolution.

```php 
usleep((int) (1000*1000*pow(1.2, $this->tokenFileReadAttempts)));
```
(Same goes for the same change in other places where this was changed.)

allemas replied:
I do not know if you have seen but in the file aws-sdk-php/src/Credentials/InstanceProfileProvider.php there is the same fix to do in line 262 : 

```
    private function handleRetryableException(
        \Exception $e,
        $retryOptions,
        $message
    ) {
        $isRetryable = true;
        if (!empty($status = $this->getExceptionStatusCode($e))
            && isset($retryOptions['blacklist'])
            && in_array($status, $retryOptions['blacklist'])
        ) {
            $isRetryable = false;
        }
        if ($isRetryable && $this->attempts < $this->retries) {
            sleep((int) pow(1.2, $this->attempts));            <-----
        } else {
            throw new CredentialsException($message);
        }
    }
```

Zombaya commented on tests/CloudFront/fixtures/test2.pem:
```diff
@@ -0,0 +1,27 @@
+-----BEGIN RSA PRIVATE KEY-----
```
I've replaced this key by generating a new key in memory during each test.

Zombaya commented on .changes/nextrelease/php8:
```diff
@@ -0,0 +1,7 @@
+    {
+        "type": "enhancement",
+        "category": "",
+        "description": "Add support for php 8.x"
```
I've replaced it by your suggestion.

Zombaya commented:
@stobrien89 , I fixed all your latest remarks and rebased everything on master once more.

However, I'm running into a new issue with the CI regarding the test [`Aws\Test\CloudFront\SignerTest::testBadPrivateKeyPath()`](https://github.com/Zombaya/aws-sdk-php/runs/6467922428?check_suite_focus=true).  

It's throwing some sort of error related to openssl, but I'm not being to reproduce the error locally.  Nor the underlying code nor the testcode have changed fundamentally with this PR, so I'm a bit out my depth here on how to fix this.

divinity76 commented:
PS aws-sdk-php seems to be affected by https://github.com/PHPCompatibility/PHPCompatibility/issues/1339

stobrien89 commented:
Hi @Zombaya,

So sorry for the delayed response— I missed your last message.  I'll take another look.  Thanks again for your patience!

stobrien89 commented on src/Api/Serializer/JsonBody.php:
```diff
@@ -29,7 +29,7 @@ public function __construct(Service $api)
     {
         return 'application/x-amz-json-'
-            . number_format($service->getMetadata('jsonVersion'), 1);
+            . number_format(($version = $service->getMetadata('jsonVersion')) !== null ? $version : 0.0, 1);
```
Could you make this multiple lines?

stobrien89 commented on tests/Polyfill/PHPUnit9/PHPUnitCompatTrait.php:
```diff
@@ -0,0 +1,130 @@
+
+    public function assertArraySubset($subset, $array, $checkForObjectIdentity = false, $message = '')
+    {
+        if ( ! (\is_array($subset) || $subset instanceof ArrayAccess)) {
```
Can you change the spacing here and for the line below to `!(\is_array...`?

stobrien89 commented on tests/bootstrap.php:
```diff
@@ -3,17 +3,21 @@
+$loader = require __DIR__.'/../vendor/autoload.php';
 
-if (!class_exists('\PHPUnit\Framework\Constraint\RegularExpression')) {
+if ( ! defined('PHPUNIT_COMPOSER_INSTALL')) {
```
Can you change the not operator spacing in this file as well? i.e. (!class_exists...

stobrien89 commented on .github/workflows/tests.yml:
```diff
@@ -76,6 +90,6 @@ jobs:
 
     #generate code coverage
-    - if: ${{ matrix.php-versions == '7.1' && matrix.composer-options == '' }}
+    - if: ${{ (matrix.php-versions == '7.1' || matrix.php-versions == '8.0') && matrix.composer-options == '' }}
```
Let's just upgrade the version here if possible.

stobrien89 reviewed (changes requested):
Left a few comments, but I think we're getting close :)

Zombaya commented on src/S3/S3Client.php:
```diff
@@ -429,7 +429,7 @@ public function __construct(array $args)
      */
     public static function isBucketDnsCompatible($bucket)
```
I've added a type check and some more cases to the unittest.

Zombaya commented on tests/CloudFront/UrlSignerTest.php:
```diff
@@ -28,30 +31,28 @@ public function testCreatesUrlSignersForHttp()
         );
         $urlObject = new Uri($url);
-        $query = \GuzzleHttp\Psr7\Query::parse($urlObject->getQuery());
+        $query     = \GuzzleHttp\Psr7\Query::parse($urlObject->getQuery());
```
I assume @27pchrisl either configured his IDE to do this automatically or has the same preference as me, which is aligned `=`-signs :slightly_smiling_face: .

I've removed it for now and reverted to what it was.

Zombaya commented:
I fixed all final remarks you had and rebased everything on master again.

I think the formatting-issues could have been avoided if there was a php-cs-fixer-configuration-file available and a lint-check in github-CI.  

I'm trying out some things to see if I can make a valuable contribution for this in [Zombaya/aws-sdk-php#4 php-cs-fixer](https://github.com/Zombaya/aws-sdk-php/pull/4). 
Please let me know if you think this would not be helpful so I avoid spending too much time on it.

Zombaya commented:
I've rebased onto master again. 

I needed to modify `Aws\Test\S3\StreamWrapperV2ExistenceTest` to use the new `PhpUnitCompatTrait` and use the methods for expecting errors instead of the annotations.

Zombaya commented on tests/S3/StreamWrapperV2ExistenceTest.php:
```diff
@@ -9,6 +9,7 @@
 use Aws\S3\Exception\S3Exception;
 use Aws\S3\S3Client;
 use Aws\S3\StreamWrapper;
+use Aws\Test\Polyfill\PHPUnit\PHPUnitCompatTrait;
```
This file was updated in my last rebase.

Zombaya commented on tests/EndpointParameterMiddlewareTest.php:
```diff
@@ -90,7 +90,7 @@ public function testCorrectlyOutputsHost(
 
         $handler = $list->resolve();
-        $handler($command, new Request('POST', $endpoint));
+        $handler($command, new Request('POST', $endpoint))->wait(false);
```
I found out why this was probably added.  

When running this test using PHP 8.0 and PHPUnit 9.x, this test is marked as a risky test because it does not run any assertions.  When adding `->wait(false)`, this disappears and the assertions are actually being done.

Zombaya commented:
@stobrien89 , just wanted to let you know I'll not be available in the coming weeks, so will not be able to help with any rebasing. 

I'm very happy with the work on #2456 and will be implementing the result internally as soon as possible :slightly_smiling_face:.  Nice work:+1:!

SamRemis commented on src/S3/S3Client.php:
```diff
@@ -429,7 +429,11 @@ public function __construct(array $args)
+            return false;
+        }
+
+        $bucketLen = $bucket !== null ? strlen($bucket) : 0;
```
This check is not necessary if we are checking for is_string above; it will never be null and always be a string, right?

SamRemis commented on tests/Polyfill/PHPUnit4/PHPUnitCompatTrait.php:
```diff
@@ -0,0 +1,141 @@
+<?php
+
```
Is there a reason we would use our own Polyfill instead of taking a dependency on something widely tested and maintained like https://github.com/Yoast/PHPUnit-Polyfills?

SamRemis replied:
I'd think it's worth having as little code to maintain as possible by using the polyfill, especially when there's a dependency that does exactly that.  I'd prefer it, yes.  @stobrien89, do you agree? 

Also as a side note, sorry this is taking so long.  It's a huge PR and carefully looking at every single file over and over again is very time consuming.  Thank you for being patient, and I'll do my best to prioritize it higher.  If you don't see me respond for a bit, feel free to @ me to give me a notification.  That said, I'll try to look at this every day until it's merged.

stobrien89 replied:
@SamRemis @27pchrisl,

I definitely agree.  Especially with us looking forward to the next major version, having as little code to maintain as possible will be ideal.  Thanks again for all of the work on this @Zombaya and @27pchrisl!

SamRemis commented on tests/CloudFront/UrlSignerTest.php:
```diff
@@ -28,30 +31,28 @@ public function testCreatesUrlSignersForHttp()
-        $ts = time() + 1000;
-        $key = $_SERVER['CF_PRIVATE_KEY'];
-        $kp = $_SERVER['CF_KEY_PAIR_ID'];
+        $ts     = time() + 1000;
```
spacing

SamRemis commented on tests/CloudFront/UrlSignerTest.php:
```diff
@@ -61,23 +62,21 @@ public function testCreatesUrlSignersWithSpecialCharacters()
-        $ts = time() + 1000;
-        $key = $_SERVER['CF_PRIVATE_KEY'];
-        $kp = $_SERVER['CF_KEY_PAIR_ID'];
+        $ts     = time() + 1000;
```
spacing

SamRemis commented on tests/CloudFront/UrlSignerTest.php:
```diff
@@ -61,23 +62,21 @@ public function testCreatesUrlSignersWithSpecialCharacters()
 
         $invalidUri = 'http://abc.cloudfront.net/images/éüàçµñåœŒ.jpg?query key=query value';
-        $uri = new Uri($invalidUri);
+        $uri        = new Uri($invalidUri);
```
spacing

SamRemis commented on tests/CloudFront/UrlSignerTest.php:
```diff
@@ -91,14 +90,14 @@ public function testCreatesUrlSignersWithCustomPolicy()
-            'policy' => '{}',
-            'private_key' => $_SERVER['CF_PRIVATE_KEY'],
-            'key_pair_id' => $_SERVER['CF_KEY_PAIR_ID']
+        $url    = $client->getSignedUrl(array(
```
spacing (next three lines)

SamRemis commented on tests/ClientResolverTest.php:
```diff
@@ -222,19 +221,17 @@ public function testValidatesNotInvokeStringCallable()
+        $this->assertIsCallable($callableFunction);
         $this->assertSame(
-            '\Aws\test\ClientResolverTest::checkCallable',
+            '\Aws\Test\ClientResolverTest::checkCallable',
```
Isn't this lowercase?  It should work either way, but why change it?

stobrien89 commented:
@Zombaya, @27pchrisl,

I've rebased and updated tests for `RemoveUnusedServices` and #2459.  I've also fixed some of the warnings/risky tests that appear in 7.3+ and addressed @SamRemis's most recent comments.  I've attempted to push, but it looks like I'll need to be granted access.  Thanks again!

Zombaya commented:
@stobrien89 , I've added you and the other contributors here as contributor to the project so you should be able to push to the branch.

Otherwise a second PR can be created to replace this one from your branch.

RobinHoutevelts commented:
@stobrien89 Did you manage to get the work you did above (https://github.com/aws/aws-sdk-php/pull/2403#issuecomment-1191906911) into this PR?

stobrien89 commented:
Hi @Zombaya, @27pchrisl, @RobinHoutevelts,

Apologies for the delay.  Just pushed up the changes I mentioned in my last comment, including some changes related to https://github.com/aws/aws-sdk-php/pull/2498.  

From here, I think the next step would be to take a dependency on https://github.com/Yoast/PHPUnit-Polyfills.  @SamRemis and I have discussed and with the next major version on the horizon, we'd rather not have another component to maintain.  Thanks!

Zombaya commented:
@stobrien89, @SamRemis,  I've added a dependency on `yoast/phpunit-polyfills` and `dms/phpunit-arraysubset-asserts`.

`dms/phpunit-arraysubset-asserts` was added since the removed `$this->assertArraySubset()` is used in some places and is not provided by `yoast/phpunit-polyfills`. It is [directly recommended](https://github.com/Yoast/PHPUnit-Polyfills#removed-functionality-without-phpunit-native-replacement) by `yoast/phpunit-polyfills` though.

# Strategy
* Classes which use no deprecated PHPUnit-assertions extend `PHPUnit\Framework\TestCase`
* Classes which use a deprecated PHPUnit-assertion provided by `yoast/phpunit-polyfills` extend `Yoast\PHPUnitPolyfills\TestCases\TestCase`
* Traits which use deprecated PHPUnit-assertions use the specific Trait provided by yoast which provide the used assertion
* Classes which use the deprecated method `assertArraySubset()` use the Trait provided by `dms/phpunit-arraysubset-asserts`

# Base TestCase
I opted to only use `Yoast\PHPUnitPolyfills\TestCases\TestCase` as a base-class for testcases which actually use deprecated methods.  I think this makes it more clear which testcases which are more interesting to refactor or to check in case the support for php 5.5 and PHPUnit 5 is dropped.

The other option would be that every test extends this TestCase. This way you do not need to be very aware of what methods are deprecated when writing new TestCases.

I felt that the first option was preferable.

# Main changes

## Different methodnames for setUp(), tearDown(), ...
The library by Yoast provides different methodnames for fixtures than what was used in the `PhpUnitCompatTrait`.

```diff
-- function _setUp() {}
++ function set_up() {}

-- function _tearDown() {}
++ function tear_down() {}

-- public static function _setUpBeforeClass()
++ public static function set_up_before_class()

-- public static function _tearDownAfterClass()
++ public static function tear_down_after_class()
```

## $this->readAttribute() replaced by $this->getPropertyValue()
`readAttribute()` was part of `PhpUnitCompatTrait`.  This has been replaced by [the method provided by Yoast](https://github.com/Yoast/PHPUnit-Polyfills#yoastphpunitpolyfillshelpersassertattributehelper).
```diff
-- $this->assertArrayHasKey('s3', $this->readAttribute($p, 'manifest'));
++ $this->assertArrayHasKey('s3', $this->getPropertyValue($p, 'manifest'));
```

## Replaced $this->expectNotToPerformAssertions() by $this->addToAssertionCount(1)
`expectNotToPerformAssertions()` does not exist in older versions of PHPUnit. It is also not provided by `yoast/phpunit-polyfills`.
That's why I opted to use `$this->addToAssertionCount(1)`, to avoid problems with warnings of tests with no assertions.

I've left a comment in place to eventually replace it again with `expectNotToPerformAssertions()`.

Other options would be writing a new compatibilityTrait inside our library, try to get a PR merged with `yoast/phpunit-polyfills` so
that it provides the method or rewriting the tests so that they always do an assertion.

```diff
-- $this->expectNotToPerformAssertions();
++ $this->addToAssertionCount(1); // To be replaced with $this->expectNotToPerformAssertions();
```

I could not find a lot of documentation on the method.  I couldn't find it in the docs for example. Some documentation I found:

* [YouTrack jetbrains WI-66119](https://youtrack.jetbrains.com/issue/WI-66119/New-inspection-PhpUnit-test-did-not-perform-any-assertions)
* [Stackoverflow: PHPUnit passing a test with no assertions within config](https://stackoverflow.com/questions/6666242/phpunit-passing-a-test-with-no-assertions-within-config)
* [Github PHPUnit PR#2898 Added expectNotToPerformAssertions method](https://github.com/sebastianbergmann/phpunit/pull/2898)

devicenull commented:
Was any further progress made here?  We've started to run into issues with PHP 8.1

RobinHoutevelts commented:
> Also as a side note, sorry this is taking so long. It's a huge PR and carefully looking at every single file over and over again is very time consuming. Thank you for being patient, and I'll do my best to prioritize it higher. **If you don't see me respond for a bit, feel free to @ me to give me a notification.** That said, I'll try to look at this every day until it's merged.

@SamRemis , consider this a friendly notification 👼 

Thanks everyone for all the hard work so far!

SamRemis commented:
@RobinHoutevelts thanks for the ping :) since the last update on Tuesday or Wednesday I've been working through this about 40 files a day, I'm gonna try to finish it up tomorrow or monday.  Only issue I've found so far is a missing space, which I won't block on, so it's getting close :)

SamRemis commented:
Just to update, I have about 20 files left to go through, which I won't be able to do before today's release, but I should be able to finish today.  Assuming I find no issues, which I haven't yet in the first 140 files, I will be able to merge tonight and this should go out with tomorrow's release.

roborourke commented:
Absolute legend @Zombaya

Zombaya commented:
@SamRemis, @stobrien89, you made me one happy camper by getting this merged and released today!

Many thanks and much kudos for the effort you spent and the thorough review you did multiple times.  I was impressed with the level of detail you went over everything, as well as the respectfulness of the notes when something was not quite correct.

Small shout-out to @27pchrisl for initiating the work on this PR and continuing to help after reviews have been made. 

I think it's safe to say this was as PR me and a lot of others were all looking forward to be merged into master :slightly_smiling_face:.
